File: python/SparkTPApp3.py
#! python
# -*- coding: utf-8 -*-
import sys
import pyspark
import petasky_schema

compte = "ecoquery"


def lineToStrings(l):
    return list(map(lambda c: c.strip(), l.split(",")))


def buildTuple(r):
    oId = r[petasky_schema.s_objectId()]
    sId = r[petasky_schema.s_sourceId()]
    ra = float(r[petasky_schema.s_ra()])
    decl = float(r[petasky_schema.s_decl()])
    return (oId, 1, sId, ra, ra, decl, decl)


def aggTuples(t1, t2):
    return (
        t1[0],  # oId
        t1[1] + t2[1],  # count
        max(t1[2], t2[2]),  # max sourceId
        min(t1[3], t2[3]),  # min ra
        max(t1[4], t2[4]),  # max ra
        min(t1[5], t2[5]),  # min decl
        max(t1[6], t2[6]),  # max decl
    )


def aggregateByObjectId(inputDir, sc):
    rows = sc.textFile(inputDir).map(lineToStrings)
    noNulls = rows.filter(lambda r: r[petasky_schema.s_objectId()] != "NULL")
    filteredCols = noNulls.map(buildTuple)
    keyed = filteredCols.map(lambda t: (t[0], t))
    agg = keyed.reduceByKey(aggTuples)
    return agg.map(lambda p: p[1])  # extract the resulting tuples


if __name__ == "__main__":
    args = sys.argv
    if len(args) > 2 and "-h" not in args and "--help" not in args:
        sc = pyspark.SparkContext(appName="SparkTPApp3-" + compte)
        result = aggregateByObjectId(args[1], sc)
        result.map(lambda t: ",".join(map(str, t))).saveAsTextFile(args[2])
    else:
        print(
            "Usage: spark-submit --py-files petasky_schema.py "
            + "--files Source.sql,Object.sql "
            + args[0]
            + " hdfs:///user/"
            + compte
            + "/repertoire-donnees"
            + " hdfs:///user/"
            + compte
            + "/repertoire-resultat"
        )

File: src/main/scala/SparkTPApp3.scala
import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}
import schema.PetaSkySchema
import Math.{min, max}

object SparkTPApp3 {
  val compte = "ecoquery" // changer avec votre login

  // Version avec une case class
  case class AP3Agg(objectId: String, count: Long, lastSrc: String,
                    minRA: Double, maxRA: Double,
                    minDecl: Double, maxDecl: Double) {
    def agg(b: AP3Agg): AP3Agg = {
      AP3Agg(objectId, count + b.count,
        if (lastSrc >= b.lastSrc) lastSrc else b.lastSrc,
        min(minRA, b.minRA), max(maxRA, b.maxRA),
        min(minDecl, b.minDecl), max(maxDecl, b.maxDecl)
      )
    }

    def toTuple: (String, Long, String, Double, Double, Double, Double) = {
      (objectId, count, lastSrc, minRA, maxRA, minDecl, maxDecl)
    }
  }

  def toAP3Agg(input: Array[String]): AP3Agg = {
    AP3Agg(input(PetaSkySchema.s_objectId), 1L, input(PetaSkySchema.s_sourceId),
      input(PetaSkySchema.s_ra).toDouble, input(PetaSkySchema.s_ra).toDouble,
      input(PetaSkySchema.s_decl).toDouble, input(PetaSkySchema.s_decl).toDouble)
  }

  def aggregateByObjectIdCC(inputDir: String, sc: SparkContext):
  RDD[(String, Long, String, Double, Double, Double, Double)] = {
    sc.textFile(inputDir)
      .map(_.split(",").map(_.trim)) // transformer chaque ligne en un tableau de String)
      .map(toAP3Agg)
      .filter(_.objectId != "NULL")
      .map(a => (a.objectId, a))
      .reduceByKey((a, b) => a.agg(b))
      .map(r => r._2.toTuple)
  }

  // Version purement tuple
  def aggregateByObjectIdT(inputDir: String, sc: SparkContext):
  RDD[(String, Long, String, Double, Double, Double, Double)] = {
    sc.textFile(inputDir)
      .map(_.split(",").map(_.trim)) // transformer chaque ligne en un tableau de String
      .filter(l => l(PetaSkySchema.s_objectId) != "NULL")
      .map(l => (l(PetaSkySchema.s_objectId), 1L, l(PetaSkySchema.s_sourceId),
        l(PetaSkySchema.s_ra).toDouble, l(PetaSkySchema.s_ra).toDouble,
        l(PetaSkySchema.s_decl).toDouble, l(PetaSkySchema.s_decl).toDouble)) // on extrait les attribus d'intérêt
      .map(l => (l._1, l))
      .reduceByKey((a, b) =>
        (a._1,
          a._2 + b._2,
          if (a._3 >= b._3) a._3 else b._3,
          min(a._4, b._4), max(a._5, b._5),
          min(a._6, b._6), max(a._7, b._7)))
      .map(r => r._2)
  }

  def main(args: Array[String]): Unit = {
    if (args.length > 2) {
      val conf = new SparkConf().setAppName("SparkTPApp3-" + compte)
      val sc = new SparkContext(conf)
      val result =
        if ("--tuple" == args(0))
          aggregateByObjectIdT(args(1), sc)
        else
          aggregateByObjectIdCC(args(1), sc)
      result
        .map(r => "%s,%d,%s,%s,%s,%s,%s".format(r._1, r._2, r._3, r._4, r._5, r._6, r._7))
        .saveAsTextFile(args(2))

    } else {
      println("Usage: spark-submit --class SparkTPApp3 /home/" + compte + "/SparkTPApp-correction-assembly-1.0.jar " +
        "[ --caseClass | --tuple ] " +
        "hdfs:///user/" + compte + "/repertoire-donnees " +
        "hdfs:///user/" + compte + "/repertoire-resultat")
    }
  }
}
